cosmwasm/lib/Makefile

```diff
@@ -0,0 +1,172 @@
+
+RustEnclave_C_Files := $(wildcard ./wasmi-runtime/*.c)
+RustEnclave_C_Objects := $(RustEnclave_C_Files:.c=.o)
+RustEnclave_Include_Paths := -I$(CUSTOM_COMMON_PATH)/inc -I$(CUSTOM_EDL_PATH) -I$(SGX_SDK)/include -I$(SGX_SDK)/include/tlibc -I$(SGX_SDK)/include/stlport -I$(SGX_SDK)/include/epid -I ./wasmi-runtime -I./include
+
+RustEnclave_Link_Libs := -L$(CUSTOM_LIBRARY_PATH) -lcompiler-rt-patch -lenclave
+RustEnclave_Compile_Flags := $(SGX_COMMON_CFLAGS) -nostdinc -fvisibility=hidden -fpie -fstack-protector $(RustEnclave_Include_Paths)
+RustEnclave_Link_Flags := $(SGX_COMMON_CFLAGS) -Wl,--no-undefined -nostdlib -nodefaultlibs -nostartfiles -L$(SGX_LIBRARY_PATH) \
+	-Wl,--whole-archive -l$(Trts_Library_Name) -Wl,--no-whole-archive \
+	-Wl,--start-group -lsgx_tstdc -l$(Service_Library_Name) -l$(Crypto_Library_Name) $(RustEnclave_Link_Libs) -Wl,--end-group \
+	-Wl,-Bstatic -Wl,-Bsymbolic -Wl,--no-undefined \
+	-Wl,-pie,-eenclave_entry -Wl,--export-dynamic  \
+	-Wl,--defsym,__ImageBase=0 \
+	-Wl,--gc-sections \
+	-Wl,--version-script=wasmi-runtime/Enclave.lds
+
+RustEnclave_Name := wasmi-runtime/enclave.so
+Signed_RustEnclave_Name := bin/enclave.signed.so
+
+.PHONY: all
+all: enclave-ffi-types $(App_Name) $(Signed_RustEnclave_Name)
+
+.PHONY: enclave-ffi-types
+enclave-ffi-types:
+	cd enclave-ffi-types && cargo build
+	mkdir -p ./sgx-vm/target/headers
+	mkdir -p ./wasmi-runtime/target/headers
+	cp -f ./enclave-ffi-types/target/headers/enclave-ffi-types.h ./wasmi-runtime/target/headers
+	cp -f ./enclave-ffi-types/target/headers/enclave-ffi-types.h ./sgx-vm/target/headers
+
+######## EDL Objects ########
+
+$(Enclave_EDL_Files): $(SGX_EDGER8R) wasmi-runtime/Enclave.edl
+	$(SGX_EDGER8R) --trusted wasmi-runtime/Enclave.edl --search-path $(SGX_SDK)/include --search-path incubator-teaclave-sgx-sdk/edl --trusted-dir wasmi-runtime
+	$(SGX_EDGER8R) --untrusted wasmi-runtime/Enclave.edl --search-path $(SGX_SDK)/include --search-path incubator-teaclave-sgx-sdk/edl --untrusted-dir sgx-vm
+	@echo "GEN  =>  $(Enclave_EDL_Files)"
+
+######## App Objects ########
+
+sgx-vm/Enclave_u.o: $(Enclave_EDL_Files)
+	@$(CC) $(App_C_Flags) -c sgx-vm/Enclave_u.c -o $@
+	@echo "CC   <=  $<"
+
+$(App_Enclave_u_Object): sgx-vm/Enclave_u.o
+	$(AR) rcsD $@ $^
+	cp $(App_Enclave_u_Object) ./lib
+
+$(App_Name): enclave-ffi-types $(App_Enclave_u_Object) $(App_SRC_Files)
+	@cd sgx-vm && SGX_SDK=$(SGX_SDK) cargo build $(App_Rust_Flags)
+	@echo "Cargo  =>  $@"
+	mkdir -p bin
+	cp $(App_Rust_Path)/sgx-vm ./bin
+
+######## Enclave Objects ########
+
+wasmi-runtime/Enclave_t.o: $(Enclave_EDL_Files)
+	@$(CC) $(RustEnclave_Compile_Flags) -c wasmi-runtime/Enclave_t.c -o $@
+	@echo "CC   <=  $<"
+
+$(RustEnclave_Name): enclave compiler-rt wasmi-runtime/Enclave_t.o
+	cp incubator-teaclave-sgx-sdk/compiler-rt/libcompiler-rt-patch.a ./lib
+	@$(CXX) wasmi-runtime/Enclave_t.o -o $@ $(RustEnclave_Link_Flags)
+	@echo "LINK =>  $@"
+
+$(Signed_RustEnclave_Name): enclave-ffi-types $(RustEnclave_Name)
```

Comment:
I believe that you don't need the `enclave-ffi-types` rule here at all, since the `$(RustEnclave_Name)` rule runs tthe `enclave` rule, which runs cargo/xargo before running anything else, and that should place the headers in their right place.

Reply:
Yep but we get an error about missing `enclave-ffi-types.h` in those directories so we added this rule for now

Reply:
can you please show me the error when building only this rule?

Reply:
resolved in call